src/core/comm/coll.cc

```diff
@@ -0,0 +1,389 @@
+  if (nb_comms != nullptr) { MAX_NB_COMMS = atoi(nb_comms); }
+  assert(MAX_NB_COMMS > 0);
+#ifdef LEGATE_USE_GASNET
+  int provided, init_flag = 0;
+  CHECK_MPI(MPI_Initialized(&init_flag));
+  if (!init_flag) {
+    CHECK_MPI(MPI_Init_thread(&argc, &argv, MPI_THREAD_MULTIPLE, &provided));
+  } else {
+    log_coll.print(
+      "Warning: MPI has been initialized by others, make sure MPI is initialized with "
+      "MPI_THREAD_MULTIPLE");
+  }
+  mpi_comms.resize(MAX_NB_COMMS, MPI_COMM_NULL);
+  for (int i = 0; i < MAX_NB_COMMS; i++) { CHECK_MPI(MPI_Comm_dup(MPI_COMM_WORLD, &mpi_comms[i])); }
+#else
+  thread_comms.resize(MAX_NB_COMMS);
+  for (int i = 0; i < MAX_NB_COMMS; i++) {
+    thread_comms[i].ready_flag = false;
+    thread_comms[i].buffers    = nullptr;
+    thread_comms[i].displs     = nullptr;
+  }
+#endif
+  coll_inited = true;
+  return CollSuccess;
+}
+
+int collFinalize(void)
+{
+  assert(coll_inited == true);
+  coll_inited = false;
+#ifdef LEGATE_USE_GASNET
+  for (int i = 0; i < MAX_NB_COMMS; i++) { CHECK_MPI(MPI_Comm_free(&mpi_comms[i])); }
+  mpi_comms.clear();
+  return MPI_Finalize();
+#else
+  for (int i = 0; i < MAX_NB_COMMS; i++) { assert(thread_comms[i].ready_flag == false); }
+  thread_comms.clear();
+  return CollSuccess;
+#endif
+}
+
+int collGetUniqueId(int* id)
+{
+  *id = current_unique_id;
+  current_unique_id++;
+  assert(current_unique_id <= MAX_NB_COMMS);
+  return CollSuccess;
+}
+
+#ifdef LEGATE_USE_GASNET
+MPI_Datatype dtypeToMPIDtype(CollDataType dtype)
+{
+  MPI_Datatype mpi_dtype = MPI_BYTE;
+  switch (dtype) {
+    case CollDataType::CollInt8: mpi_dtype = MPI_INT8_T; break;
+    case CollDataType::CollChar: mpi_dtype = MPI_CHAR; break;
+    case CollDataType::CollUint8: mpi_dtype = MPI_UINT8_T; break;
+    case CollDataType::CollInt: mpi_dtype = MPI_INT; break;
+    case CollDataType::CollUint32: mpi_dtype = MPI_UINT32_T; break;
+    case CollDataType::CollInt64: mpi_dtype = MPI_INT64_T; break;
+    case CollDataType::CollUint64: mpi_dtype = MPI_UINT64_T; break;
+    case CollDataType::CollFloat: mpi_dtype = MPI_FLOAT; break;
+    case CollDataType::CollDouble: mpi_dtype = MPI_DOUBLE; break;
+    default: log_coll.fatal("Unknown datatype"); assert(0);
+  }
+  return mpi_dtype;
+}
+
+static inline int match2ranks(int rank1, int rank2)
+{
+  // tag: seg idx + rank_idx + tag
+  // send_tag = sendto_global_rank * 10000 + global_rank (concat 2 ranks)
+  // which dst seg it sends to (in dst rank)
+  // recv_tag = global_rank * 10000 + recvfrom_global_rank (concat 2 ranks)
+  // idx of current seg we are receving (in src/my rank)
+  // example:
+  // 00 | 01 | 02 | 03
+  // 10 | 11 | 12 | 13
+  // 20 | 21 | 22 | 23
+  // 30 | 31 | 32 | 33
+  // 01's send_tag = 10, 10's recv_tag = 10, match
+  // 12's send_tag = 21, 21's recv_tag = 21, match
+
+  int tag;
+  constexpr int const max_ranks = 10000;
+  tag                           = rank1 * max_ranks + rank2;
```

Comment:
I remember I left a comment about the encoding scheme. Any comment on that? I guess this integer arithmetic is fine, but I feel bit operators would be more natural and also more efficient.